### PR TITLE
add groups to the audit trail

### PR DESCRIPTION
**Pull request by deads2k**
Adds groups to the string that gets put in the audit log.

@soltysh @sttts

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37102)

### CODE REVIEW

File: pkg/apiserver/filters/audit.go [hunk 1]
```diff
@@ -125,6 +124,18 @@ func WithAudit(handler http.Handler, attributeGetter RequestAttributeGetter, out
...
+func auditStringSlice(inList []string) string {
+	if len(inList) == 0 {
+		return ""
+	}
+
+	quotedElements := make([]string, len(inList))
+	for i, in := range inList {
+		quotedElements[i] = fmt.Sprintf("%q", in)
+	}
+	return strings.Join(quotedElements, ", ")
```
Comment (sttts) [inline pkg/apiserver/filters/audit.go]: ```"\"" + strings.Join(inList, "\", \") + "\""``` ;-)

File: pkg/apiserver/filters/audit.go [hunk 2]
```diff
@@ -96,27 +96,26 @@ func WithAudit(handler http.Handler, attributeGetter RequestAttributeGetter, out
 			internalError(w, req, err)
 			return
 		}
+
+		groups := auditStringSlice(attribs.GetUser().GetGroups())
```
Comment (sttts) [inline pkg/apiserver/filters/audit.go]: `<none>` if len is 0.

File: pkg/apiserver/filters/audit.go (hunk 1 above)
Comment (deads2k) [inline pkg/apiserver/filters/audit.go]: > "\"" + strings.Join(inList, "\", \"") + "\"" ;-)

That's not obvious to me.  Each individual string has quotes around it already, so in the log you get `"\"group1\", \"group2\""`.  You want `"\"\"group1\"\", \"\"group2\"\""`?


For instance, `user="system:apiserver" groups="\"system:masters\""`

**Comment by soltysh** [comment]
@deads2k make sure to update accompanying `_test.go`

**Comment by soltysh** [comment]
Otherwise LGTM

File: pkg/apiserver/filters/audit.go (hunk 1 above)
Comment (sttts) [inline pkg/apiserver/filters/audit.go]: I meant this:

```golang
func auditStringSlice(inList []string) string {
  return "\"" + strings.Join(inList, "\", \"") + "\""
}
```

Comment (deads2k) [inline pkg/apiserver/filters/audit.go]: > But, my other comment is more important: we should get rid of the space between the groups.

will remove the space.

File: pkg/apiserver/filters/audit.go [hunk 3]
```diff
@@ -96,27 +96,26 @@ func WithAudit(handler http.Handler, attributeGetter RequestAttributeGetter, out
...
 		}
 		namespace := attribs.GetNamespace()
 		if len(namespace) == 0 {
 			namespace = "<none>"
 		}
 		id := uuid.NewRandom().String()
 
-		line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q as=%q asgroups=%q namespace=%q uri=%q\n",
-			time.Now().Format(time.RFC3339Nano), id, utilnet.GetClientIP(req), req.Method, attribs.GetUser().GetName(), asuser, asgroups, namespace, req.URL)
+		line := fmt.Sprintf("%s AUDIT: id=%q ip=%q method=%q user=%q groups=%q as=%q asgroups=%q namespace=%q uri=%q\n",
```
Comment (deads2k) [inline pkg/apiserver/filters/audit.go]: > Which means that with foo"bar as the only group you will get "\"foo\\\"bar\"" as output?

That is the result I expect.  One parsing rule.



> Compare also the test-case below: would have expectedgroups=instead ofgroups=""`.

Also the expected result.  It's a quoted string of quoted strings separated by spaces.  Same as `asgroups` before it.

**Comment by deads2k** [comment]
Updated for `"<none>"`

File: pkg/apiserver/filters/audit.go (hunk 3 above)
Comment (sttts) [inline pkg/apiserver/filters/audit.go]: I don't want to see people using sed or awk to parse these strings :scream:

Comment (soltysh) [inline pkg/apiserver/filters/audit.go]: That's why we need to write that advanced audit 😉

**Pull request merged**